Agpasswordreset

teravest opened the pull request:
This should be pretty and working now.
Links now appear on index.psp

teravest committed:
First attempt at automatic password reset

It prints the email message if cannot send an email
so you can copy the link into your browser when using the
virtual machine.

Right now for some reason I have to wait about 20-30 seconds
before the link with the code to change the password is valid
I'm not sure what that is about, I could add a sleep statement if
we think it will be a problem.

teravest committed:
This now has the american gut aesthetic

There are links on index.psp now

teravest committed:
cleaned up comments and removed commented out code

adamrp commented:
I'm going to deploy this to webdev so that we can test the emailing functionality of this feature.

ElDeveloper commented:
Owe you :beer: @teravest!!

On Jan 31, 2014, at 4:16 PM, teravest notifications@github.com wrote:

> This should be pretty and working now.
> Links now appear on index.psp
> 
> You can merge this Pull Request by running
> 
>   git pull https://github.com/teravest/qiime_web_app agpasswordreset
> Or view, comment on, or merge it at:
> 
>   https://github.com/qiime/qiime_web_app/pull/607
> 
> Commit Summary
> 
> First attempt at automatic password reset
> This now has the american gut aesthetic
> cleaned up comments and removed commented out code
> File Changes
> 
> M python_code/data_access/ag_data_access.py (54)
> A www/americangut/forgot_password.psp (231)
> M www/americangut/index.psp (5)
> A www/americangut/retrieve_kitid.psp (108)
> Patch Links:
> 
> https://github.com/qiime/qiime_web_app/pull/607.patch
> https://github.com/qiime/qiime_web_app/pull/607.diff
> —
> Reply to this email directly or view it on GitHub.

teravest commented:
I expect that :beer: to be giving to me during Stout month!

ElDeveloper commented:
You got it!!!

On Jan 31, 2014, at 4:20 PM, teravest notifications@github.com wrote:

> I expect that  to be giving to me during Stout month!
> 
> —
> Reply to this email directly or view it on GitHub.

teravest commented:
Oh yeah, you can view the new stored procedures in SQL developer

adamrp commented:
I clicked the "Forgot Kit ID" link, and when I enter an incorrect email address, everything looks great.  If I put in a real one, I am sent an email but the page goes blank -- maybe add a message like "Check your inbox" or something?  Also, the email I received begins with:

Your American Gut Kit ID is %s

So I think the string formatting is missing something :)

ElDeveloper commented:
I know we don't have consistency between what's deployed and what's in the repository, but would you mind adding those procedures to the `database/procedures/` folder? 

On Jan 31, 2014, at 4:21 PM, teravest notifications@github.com wrote:

> Oh yeah, you can view the new stored procedures in SQL developer
> 
> —
> Reply to this email directly or view it on GitHub.

adamrp commented on www/americangut/forgot_password.psp:
```diff
@@ -0,0 +1,231 @@
+        
+        MESSAGE=('I reset the password on American Gut Kit ID %s please click '
+                 'the link below within two hours\n'
+                 'http://mircobio.me/americangut/forgot_password.psp?email=%s;'
```
Typo here: should be microbio.me

adamrp replied:
Also, this needs to be scrubbed; email addresses will contain an @ symbol almost invariably, and that needs to be converted to it's URL-safe equivalent (%40).  Values passed via GET request are separated by ampersands instead of semicolons.  A URL that looked like this worked for me:

```
http://webdev.microbio.me/americangut/forgot_password.psp?email=adam.robbinspianka%40colorado.edu&kit_id=AG_beta_9&passcode=5tYy8PLTAjmI1BPkIhL7
```

adamrp replied:
ehh, damn git.  like THIS:

http://webdev.microbio.me/americangut/forgot_password.psp?email=adam.robbinspianka%40colorado.edu&kit_id=AG_beta_9&passcode=5tYy8PLTAjmI1BPkIhL7

ElDeveloper replied:
Using `urllib.parse.quote` should do the trick.

adamrp commented:
Got a modpython traceback (originating from a KeyError for 'email') when submitting forgot_password.psp

teravest commented:
I'll fix all these on Monday morning. :ski: time

teravest committed:
Addressed comments form @adamrp and @eldeveloper

My only other thought it should we only do this after the kit
has been verified?

teravest committed:
attempt to fix modpython traceback on emailß

teravest committed:
2nd attempt at modpython traceback

teravest commented:
All comments should be addressed now.  
The only other question is do we want to all people to change a password if the kit has not been verified?   I don't think we should. Any other opinions?

ElDeveloper commented:
This would lead to a case where they would not be able to verify their kit because they don't have any access to their accounts? Do we want this? :scream: .

On Feb 3, 2014, at 11:27 AM, teravest notifications@github.com wrote:

> All comments should be addressed now.
> 
> The only other question is do we want to all people to change a password if the kit has not been verified? I don't think we should. Any other opinions?
> 
> —
> Reply to this email directly or view it on GitHub.

teravest commented:
I will add the check

On Mon, Feb 3, 2014 at 10:51 AM, Yoshiki Vázquez Baeza <
notifications@github.com> wrote:

> This would lead to a case where they would not be able to verify their kit
> because they don't have any access to their accounts? Do we want this?
> :scream: .
> 
> On Feb 3, 2014, at 11:27 AM, teravest notifications@github.com wrote:
> 
> > All comments should be addressed now.
> > 
> > The only other question is do we want to all people to change a password
> > if the kit has not been verified? I don't think we should. Any other
> > opinions?
> > 
> > ## 
> > 
> > Reply to this email directly or view it on GitHub.
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/607#issuecomment-33980411
> .

teravest committed:
added changes to prevent unverified kit from

password changes.

teravest commented:
its all on webdev now.

teravest commented:
I just checked with @ackermag and she is going to make sure the IRB is ok with this today.  We should not merge until we get confirmation from Gail.

adamrp commented on www/americangut/forgot_password.psp:
```diff
@@ -0,0 +1,245 @@
+                    req.write('<p>%s</p>' % MESSAGE)
+            except:
+                req.write("<p>There was a problem sending you the password reset code. Please contact"
+                    " us directly at <a href=\"mailto:info@americangut.org\">"
```
Can you change all the mailto links to open in a new window by adding some `target` to the anchor tag?

adamrp replied:
To be clear, this comment only applies to mailto links that appear in a browser window (and not to mailto links that appear in emails we send)

adamrp commented on www/americangut/retrieve_kitid.psp:
```diff
@@ -0,0 +1,110 @@
+            " us directly at <a href=\"mailto:info@americangut.org\">"
+            "info@americangut.org</a>.</p>")
+
+        #remove the line below before merging
```
I guess this line and the line below should be removed? hehe

adamrp commented on database/procedures/ag_is_kit_verified.sql:
```diff
@@ -0,0 +1,18 @@
+create or replace PROCEDURE AG_GET_KIT_ID_BY_EMAIL(
```
This file appears to be identical to database/procedures/ag_get_kit_id_by_email.sql

adamrp commented on www/americangut/forgot_password.psp:
```diff
@@ -0,0 +1,245 @@
+            # add new pass to the database
+            updated_data = agDataAccess.ag_set_pass_change_code( form['email'], form['kit_id'], new_act_code )
+            
+            MESSAGE=('I reset the password on American Gut Kit ID %s please click '
```
I think this should be changed to the passive voice ("The password on AG Kit ID XXXX has been reset").

adamrp commented:
After I submit a new password, the link in the email remains active.  I think it should be inactivated (probably just by changing the passcode).

adamrp commented:
Also, after password change, a lot of sites send an email to the person saying "Your password on our site has been changed.  If you did not authorize this change, please contact us immediately at XXXX"

adamrp commented:
Oh!  I did get an email -- sorry, commented too quickly :)

adamrp commented on www/americangut/forgot_password.psp:
```diff
@@ -0,0 +1,245 @@
+            try:
+                if can_send_mail():
+                    send_email(MESSAGE, 'American Gut Password Reset', form["email"])
+                    req.write('<p>You will receive an email shortly with instructions'\
```
Need a space after "instructions" on this line, or a space before "to" on the next line

adamrp replied:
Also, please remove line continuation `\` characters

adamrp commented on www/americangut/forgot_password.psp:
```diff
@@ -0,0 +1,245 @@
+            #can only change password if kit is verified
+            alphabet = "ABCDEFGHIJKLMNOPQRSTUZWXYZ"
+            alphabet += alphabet.lower()
+            alphabet += "01234567890"
```
This works, but even cooler is to import this stuff from python's `string` module at the top of the file:

``` python
from string import letters, digits
```

then use it here as:

``` python
alphabet = letters + digits
```

adamrp commented on database/procedures/ag_is_kit_verified.sql:
```diff
@@ -0,0 +1,18 @@
+END AG_GET_KIT_ID_BY_EMAIL;
+
+/*variable user_data_ REFCURSOR;
+execute ag_get_kits_by_login('ejteravest@gmail.com', :user_data_);
```
please update this -- should be `execute ag_get_kit_id_by_email('ejteravest@gmail.com', :user_data_);`

adamrp commented on python_code/data_access/ag_data_access.py:
```diff
@@ -579,3 +579,70 @@ def updateAKB(self, barcode, moldy, overloaded, other, other_text, date_of_last_
+                    return True
+            else: 
+                    return False
+        except:
```
Is there a specific exception that you can catch here?  Also, the try block should be limited to the smallest possible amount of code that you expect might throw an exception.

adamrp commented on www/americangut/forgot_password.psp:
```diff
@@ -0,0 +1,245 @@
+<hr/>
+
+<form target="_top" id="send_me_home" name="send_me_home" action="fusebox.psp" method="post">
+    <input type="hidden" id="page" name="page" value="select_task.psp" />
```
I understand this was probably copied from elsewhere, but there is no select_task.psp in the american gut project

adamrp commented on www/americangut/forgot_password.psp:
```diff
@@ -0,0 +1,245 @@
+    if new_password != confirm_password:
+        req.write('Passwords did not match, please try again')
+    else:
+        email = form['email'].replace('%40', '@')
```
suggest using `urllib.unquote` here

adamrp commented on www/americangut/forgot_password.psp:
```diff
@@ -0,0 +1,245 @@
+elif form.has_key('passcode') and form.has_key('kit_id') and form.has_key('email'):
+    #if passcode is in the form we are resetting from a forgotten password
+    kit_id = form['kit_id']
+    email = form['email'].replace('@','%40')
```
suggest using `urllib.quote`

adamrp commented on www/americangut/forgot_password.psp:
```diff
@@ -0,0 +1,245 @@
+            MESSAGE=('I reset the password on American Gut Kit ID %s please click '
+                     'the link below within two hours\n'
+                     'http://microbio.me/americangut/forgot_password.psp?email=%s;'
+                     'kit_id=%s;passcode=%s' % (form['kit_id'], form['email'].replace('@','%40'),
```
I had no idea you could use semicolons as query separators in a URL -- pretty neat, and it lets you skip using `urllib.urlencode`!

adamrp commented on www/americangut/forgot_password.psp:
```diff
@@ -0,0 +1,245 @@
+</div>
+<%
+    else:
+        req.write('<p>Your password change code is not valid Please try again</p>')
```
Not sure if trying again is really what we want them to do :p  Maybe change to something along the lines of "Click here to go to the login page" or something (with a link)

adamrp commented on www/americangut/forgot_password.psp:
```diff
@@ -0,0 +1,245 @@
+        swabs_per_kit = kit_details['swabs_per_kit']
+        kit_verification_code = kit_details['kit_verification_code']
+        ag_kit_id = kit_details['ag_kit_id']
+        agDataAccess.updateAGKit(ag_kit_id, supplied_kit_id,
```
you added a proc and a function `ag_update_kit_password`, and here is probably where you want to call it

teravest replied:
oh :shit:  good call!   I also updated ag_update_kit_password to set the pass_reset code to null. That should address a letter comment

adamrp commented on www/americangut/forgot_password.psp:
```diff
@@ -0,0 +1,245 @@
+    email = form['email'].replace('@','%40')
+    isvalid = agDataAccess.ag_verify_kit_password_change_code(form['email'], kit_id,form['passcode'])
+    if isvalid: 
+        #req.write('<p>Please enter new password</p>')
```
Remove this commented out line

adamrp commented on www/americangut/forgot_password.psp:
```diff
@@ -0,0 +1,245 @@
+    else:
+        req.write('<p style="color:red;">This information does not match our records</p>')
+        req.write('<p>Please email <a href="mailto:info@americangut.org"> for further assistance<p>')
+%>
```
Do not need to end this block of python code here, since the next thing you do is open a block of python code

adamrp commented on www/americangut/retrieve_kitid.psp:
```diff
@@ -0,0 +1,110 @@
+    kitid = agDataAccess.getAGKitbyEmail(form["email"])
+    try:
+        if len(kitid) > 0:
+            MESSAGE=('Your American Gut Kit ID is %s. You are receiving this email '\
```
Remove line continuation characters

adamrp commented on www/americangut/retrieve_kitid.psp:
```diff
@@ -0,0 +1,110 @@
+            "info@americangut.org</a>.</p>")
+
+        #remove the line below before merging
+%>
```
Do not need to end this block of python code, since the next thing that happens is opening a new block of python code

teravest committed:
Address comments form @adamrp

changed email and error messages
cleaned up code

teravest committed:
changed to use agGetKitDetails to

check if kit is verified

teravest commented:
DO NOT MERGE

We are waiting for IRB approval.  They have been given a description of this design and are happy with it and said they will approve it. It just has to be submitted in a formal amendment. @ackermag is taking care of this.

adamrp commented:
@ackermag, is this going to be included in the next scheduled IRB amendment, or will this be its own amendment?

ackermag commented:
I would like to submit as amendment with questionnaire change.
Gail

On Wed, Feb 12, 2014 at 12:11 PM, adamrp notifications@github.com wrote:

> @ackermag https://github.com/ackermag, is this going to be included in
> the next scheduled IRB amendment, or will this be its own amendment?
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/607#issuecomment-34904366
> .

## 

Gail Ackermann
Knight Lab UCB 596
University of Colorado
Boulder

adamrp commented:
Okay.  Do you already have a list of questions for the questionnaire
change?  If so, may I see it?

On Wed, Feb 12, 2014 at 6:21 PM, ackermag notifications@github.com wrote:

> I would like to submit as amendment with questionnaire change.
> Gail
> 
> On Wed, Feb 12, 2014 at 12:11 PM, adamrp notifications@github.com wrote:
> 
> > @ackermag https://github.com/ackermag, is this going to be included in
> > the next scheduled IRB amendment, or will this be its own amendment?
> > 
> > ## 
> > 
> > Reply to this email directly or view it on GitHub<
> > https://github.com/qiime/qiime_web_app/pull/607#issuecomment-34904366>
> > .
> 
> ## 
> 
> Gail Ackermann
> Knight Lab UCB 596
> University of Colorado
> Boulder
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/607#issuecomment-34938945
> .

ackermag commented:
Hi Adam
I do not. I also need to submit the information we send to participants.
 This should be really straight forward as an amendment.
We need to ask Justine what she wants to change in the questionnaire and
perhaps some other people...I think this needs to be broadly collaborative.

We could submit the password reset wording from Emily, with the info we
send with results then look at the Questionnaire later - what would you
think is best?
Gail

On Wed, Feb 12, 2014 at 7:48 PM, adamrp notifications@github.com wrote:

> Okay. Do you already have a list of questions for the questionnaire
> change? If so, may I see it?
> 
> On Wed, Feb 12, 2014 at 6:21 PM, ackermag notifications@github.com
> wrote:
> 
> > I would like to submit as amendment with questionnaire change.
> > Gail
> > 
> > On Wed, Feb 12, 2014 at 12:11 PM, adamrp notifications@github.com
> > wrote:
> > 
> > > @ackermag https://github.com/ackermag, is this going to be included
> > > in
> > > 
> > > the next scheduled IRB amendment, or will this be its own amendment?
> > > 
> > > ## 
> > > 
> > > Reply to this email directly or view it on GitHub<
> > > https://github.com/qiime/qiime_web_app/pull/607#issuecomment-34904366>
> > > 
> > > .
> > 
> > ## 
> > 
> > Gail Ackermann
> > Knight Lab UCB 596
> > University of Colorado
> > Boulder
> > 
> > ## 
> > 
> > Reply to this email directly or view it on GitHub<
> > https://github.com/qiime/qiime_web_app/pull/607#issuecomment-34938945>
> > 
> > .
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/607#issuecomment-34943449
> .

## 

Gail Ackermann
Knight Lab UCB 596
University of Colorado
Boulder

ElDeveloper commented:
Adding @JWDebelius to the conversation.

On Feb 12, 2014, at 10:02 PM, ackermag notifications@github.com wrote:

> Hi Adam
> I do not. I also need to submit the information we send to participants.
> This should be really straight forward as an amendment.
> We need to ask Justine what she wants to change in the questionnaire and
> perhaps some other people...I think this needs to be broadly collaborative.
> 
> We could submit the password reset wording from Emily, with the info we
> send with results then look at the Questionnaire later - what would you
> think is best?
> Gail
> 
> On Wed, Feb 12, 2014 at 7:48 PM, adamrp notifications@github.com wrote:
> 
> > Okay. Do you already have a list of questions for the questionnaire
> > change? If so, may I see it?
> > 
> > On Wed, Feb 12, 2014 at 6:21 PM, ackermag notifications@github.com
> > wrote:
> > 
> > > I would like to submit as amendment with questionnaire change.
> > > Gail
> > > 
> > > On Wed, Feb 12, 2014 at 12:11 PM, adamrp notifications@github.com
> > > wrote:
> > > 
> > > > @ackermag https://github.com/ackermag, is this going to be included
> > > > in
> > > > 
> > > > the next scheduled IRB amendment, or will this be its own amendment?
> > > > 
> > > > ## 
> > > > 
> > > > Reply to this email directly or view it on GitHub<
> > > > https://github.com/qiime/qiime_web_app/pull/607#issuecomment-34904366>
> > > > 
> > > > .
> > > 
> > > ## 
> > > 
> > > Gail Ackermann
> > > Knight Lab UCB 596
> > > University of Colorado
> > > Boulder
> > > 
> > > ## 
> > > 
> > > Reply to this email directly or view it on GitHub<
> > > https://github.com/qiime/qiime_web_app/pull/607#issuecomment-34938945>
> > > 
> > > .
> > 
> > ## 
> > 
> > Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/607#issuecomment-34943449
> > .
> 
> ## 
> 
> Gail Ackermann
> Knight Lab UCB 596
> University of Colorado
> Boulder
> —
> Reply to this email directly or view it on GitHub.

jwdebelius commented:
I have a version of the suggested changes I'd like to make to the questionnaire on google docs. I'm incorporating changes from Adam, Yoshiki and Emily. Right now, they're general and not prioritized. I can send @ackermag the link or post a txt version here.

adamrp commented:
Do we have an estimate for how long it will take before we get a survey we
are all happy with?  And after we do, how long does the IRB process
typically take?  I'm asking only because it would be nice to deploy the
password reset stuff that @teravest put together as soon as possible since
it is likely to alleviate a lot of the traffic on the help account.

Note that actually deploying the new survey will take some time even after
the IRB approves the change; I assume we can keep the current version of
the survey up until we are able to deploy the new one, right?

On Thu, Feb 13, 2014 at 1:25 PM, J W Debelius notifications@github.comwrote:

> I have a version of the suggested changes I'd like to make to the
> questionnaire on google docs. I'm incorporating changes from Adam, Yoshiki
> and Emily. Right now, they're general and not prioritized. I can send
> @ackermag https://github.com/ackermag the link or post a txt version
> here.
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/607#issuecomment-35021743
> .

teravest commented:
@ackermag  just told me the IRB said we could implement this :) 
I am going to try and get this updated today so we can test it out.

teravest committed:
merge conflict

ElDeveloper commented:
That's great news!

On Apr 7, 2014, at 9:34 AM, teravest notifications@github.com wrote:

> @ackermag just told me the IRB said we could implement this :) 
> I am going to try and get this updated today so we can test it out.
> 
> —
> Reply to this email directly or view it on GitHub.

teravest commented:
This should be ready for testing and review again...  I dont know how to change the title.

adamrp commented on python_code/data_access/ag_data_access.py:
```diff
@@ -698,6 +698,58 @@ def updateAKB(self, barcode, moldy, overloaded, other, other_text, date_of_last_
+    def getAGKitbyEmail(self, email):
+        """Returns a list of kitids based on email
+
+        email is email address of participant
```
email is the email address of a login (as would be found in ag_login), and not really a participant (which would be found in ag_human_survey).  Can you just change "participant" to "login" here?

adamrp commented on python_code/data_access/ag_data_access.py:
```diff
@@ -698,6 +698,58 @@ def updateAKB(self, barcode, moldy, overloaded, other, other_text, date_of_last_
+        results = con.cursor()
+        con.cursor().callproc('ag_verify_password_change_code', [email, kitid, passcode, results])
+        isgood = results.fetchone()
+        if isgood is not None:
```
lines 745 through 751 can be replaced with

``` python
return isgood is not None and isgood[0] == 1
```

teravest committed:
address comments form @adamrp

adamrp commented:
Thanks!  I'll test on webdev when I can, probably tomorrow.

ElDeveloper commented on database/procedures/ag_update_kit_password.sql:
```diff
@@ -0,0 +1,16 @@
+ 
+
+/*
+execute AG_UPDATE_KIT_PASSWORD('test', 'test!@#$');
```
!!!!!

adamrp replied:
Ahhhh yes.  This definitely needs to change.

On Tue, Apr 8, 2014 at 10:58 AM, Yoshiki Vázquez Baeza <
notifications@github.com> wrote:

> In database/procedures/ag_update_kit_password.sql:
> 
> > @@ -0,0 +1,16 @@
> > +create or replace PROCEDURE AG_UPDATE_KIT_PASSWORD
> > +(
> > -  KIT_ID_ IN VARCHAR2
> >   +, PASS_ IN VARCHAR2
> >   +) AS
> >   +BEGIN
> > -  UPDATE AG_KIT  set kit_password = PASS_, pass_reset_code = null
> > -  where supplied_kit_id = KIT_ID_;
> > -  commit;
> >   +
> >   +END AG_UPDATE_KIT_PASSWORD;
> >   +
> >   +
> >   +/*
> >   +execute AG_UPDATE_KIT_PASSWORD('test', 'test!@#$');
> 
> !!!!!
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/607/files#r11399690
> .

teravest replied:
changed.

ElDeveloper commented on python_code/data_access/ag_data_access.py:
```diff
@@ -698,6 +698,52 @@ def updateAKB(self, barcode, moldy, overloaded, other, other_text, date_of_last_
+        results = con.cursor()
+        con.cursor().callproc('ag_verify_password_change_code', [email, kitid, passcode, results])
+        isgood = results.fetchone()
+        return isgood is not None and isgood[0] == 1
```
:+1:

ElDeveloper commented on www/americangut/forgot_password.psp:
```diff
@@ -0,0 +1,234 @@
+        if can_send_mail():
+            send_email(MESSAGE, 'American Gut Password Reset', email)
+        else:
+        #    pass
```
Can you remove this line?

ElDeveloper commented:
It's deployed on webdev, for our convenience. :+1:

teravest commented:
Should we ask other people in the lab to test this as well?
Remember the links in the emails are www not webdev
On Apr 8, 2014 11:42 AM, "Yoshiki Vázquez Baeza" notifications@github.com
wrote:

> It's deployed on webdev, for our convenience. [image: :+1:]
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/607#issuecomment-39878829
> .

ElDeveloper commented:
Ah, that's truth. I think it would be a good idea to test-run this with more people than usual but I don't know what is the best way to do that.

adamrp commented:
Just tried to reset my password, but the link in the email was not formatted properly because there are non-ASCII characters in the passcode:
![screen shot 2014-04-08 at 12 11 27 pm](https://cloud.githubusercontent.com/assets/2574407/2647321/5254093a-bf49-11e3-9c0c-7edec8ea3836.png)

teravest commented:
thats awesome.... I wonder how that happened..

On Tue, Apr 8, 2014 at 12:12 PM, adamrp notifications@github.com wrote:

> Just tried to reset my password, but the link in the email was not
> formatted properly because there are non-ASCII characters in the passcode:
> [image: screen shot 2014-04-08 at 12 11 27 pm]https://cloud.githubusercontent.com/assets/2574407/2647321/5254093a-bf49-11e3-9c0c-7edec8ea3836.png
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/607#issuecomment-39882218
> .

teravest committed:
attempt to fix nonascii

teravest committed:
add link to home in menu

adamrp commented on www/americangut/forgot_password.psp:
```diff
@@ -0,0 +1,234 @@
+    new_password = form['new_password'];
+    confirm_password = form['confirm_password']
+    if new_password != confirm_password:
+        req.write('Passwords did not match, please try again')
```
This page is a little sparse.  Can you please add a link to go back to try again?  Or, even better, display this message and then highlight the text boxes.  See new_participant.psp (and the consent checkbox) and the validateConsent javascript function for an example.  @ElDeveloper, what do you think would be a good thing to do here?

teravest replied:
I'm not sure how to do this, but another option is to make it a pop over
window on top of the map like when you register a new handout kit.  That
would look a lot nicer I bet.

On Tue, Apr 8, 2014 at 3:49 PM, adamrp notifications@github.com wrote:

> In www/americangut/forgot_password.psp:
> 
> > +
> > +<div class="content">
> > +
> > +<hr/>
> > +
> > +
> > +<br/>
> > +
> > +<%
> > +agDataAccess= data_access_factory(ServerConfig.data_access_type, 'american_gut')
> > +#if the form has new_password and confirm_password we want to change the password
> > +if form.has_key('new_password') and form.has_key('confirm_password'):
> > -    new_password = form['new_password'];
> > -    confirm_password = form['confirm_password']
> > -    if new_password != confirm_password:
> > -        req.write('Passwords did not match, please try again')
> 
> This page is a little sparse. Can you please add a link to go back to try
> again? Or, even better, display this message and then highlight the text
> boxes. See new_participant.psp (and the consent checkbox) and the
> validateConsent javascript function for an example. @ElDeveloperhttps://github.com/ElDeveloper,
> what do you think would be a good thing to do here?
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/607/files#r11413618
> .

adamrp replied:
:+1: sounds beautiful

On Tue, Apr 8, 2014 at 3:59 PM, teravest notifications@github.com wrote:

> In www/americangut/forgot_password.psp:
> 
> > +
> > +<div class="content">
> > +
> > +<hr/>
> > +
> > +
> > +<br/>
> > +
> > +<%
> > +agDataAccess= data_access_factory(ServerConfig.data_access_type, 'american_gut')
> > +#if the form has new_password and confirm_password we want to change the password
> > +if form.has_key('new_password') and form.has_key('confirm_password'):
> > -    new_password = form['new_password'];
> > -    confirm_password = form['confirm_password']
> > -    if new_password != confirm_password:
> > -        req.write('Passwords did not match, please try again')
> 
> I'm not sure how to do this, but another option is to make it a pop over
> window on top of the map like when you register a new handout kit. That
> would look a lot nicer I bet.
> ... <#145435a3c4f60db8_>
> On Tue, Apr 8, 2014 at 3:49 PM, adamrp notifications@github.com wrote:
> In www/americangut/forgot_password.psp: > + > +<div class="content"> > + >
> +<hr/> > + > + > +<br/> > + > +<% > +agDataAccess=
> data_access_factory(ServerConfig.data_access_type, 'american_gut') > +#if
> the form has new_password and confirm_password we want to change the
> password > +if form.has_key('new_password') and
> form.has_key('confirm_password'): > + new_password = form['new_password'];
> 
> > - confirm_password = form['confirm_password'] > + if new_password !=
> >   confirm_password: > + req.write('Passwords did not match, please try
> >   again') This page is a little sparse. Can you please add a link to go back
> >   to try again? Or, even better, display this message and then highlight the
> >   text boxes. See new_participant.psp (and the consent checkbox) and the
> >   validateConsent javascript function for an example. @ElDeveloperhttps://github.com/ElDeveloper
> >   https://github.com/ElDeveloper, what do you think would be a good thing
> >   to do here? -- Reply to this email directly or view it on GitHub<
> >   https://github.com/qiime/qiime_web_app/pull/607/files#r11413618> .
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/607/files#r11413978
> .

ElDeveloper replied:
:+1: that sounds good.

teravest commented:
Check it out!

![magic](https://cloud.githubusercontent.com/assets/5845974/2673552/902099ca-c0f4-11e3-9fa7-e52bf827b880.png)

adamrp commented:
Beautiful!  Does it have to contact google and get all the map markers each
time a page with a map background is loaded?  It would be cool if we could
just get that info once on index.psp and then store it for that session.

On Thu, Apr 10, 2014 at 3:10 PM, teravest notifications@github.com wrote:

> Check it out!
> 
> [image: magic]https://cloud.githubusercontent.com/assets/5845974/2673552/902099ca-c0f4-11e3-9fa7-e52bf827b880.png
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/607#issuecomment-40143394
> .

teravest commented:
I'm working on it,   right now the entire command to write the map is in a
session variable.  It still takes a second to load but it faster.

On Thu, Apr 10, 2014 at 3:29 PM, adamrp notifications@github.com wrote:

> Beautiful! Does it have to contact google and get all the map markers each
> time a page with a map background is loaded? It would be cool if we could
> just get that info once on index.psp and then store it for that session.
> 
> On Thu, Apr 10, 2014 at 3:10 PM, teravest notifications@github.com
> wrote:
> 
> > Check it out!
> > 
> > [image: magic]<
> > https://cloud.githubusercontent.com/assets/5845974/2673552/902099ca-c0f4-11e3-9fa7-e52bf827b880.png
> > 
> > ## 
> > 
> > Reply to this email directly or view it on GitHub<
> > https://github.com/qiime/qiime_web_app/pull/607#issuecomment-40143394>
> > 
> > .
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/607#issuecomment-40145327
> .

rob-knight commented:
Nice!

On Apr 10, 2014, at 3:10 PM, "teravest" <notifications@github.com<mailto:notifications@github.com>> wrote:

Check it out!

[https://cloud.githubusercontent.com/assets/5845974/2673552/902099ca-c0f4-11e3-9fa7-e52bf827b880.png]https://cloud.githubusercontent.com/assets/5845974/2673552/902099ca-c0f4-11e3-9fa7-e52bf827b880.png

—
Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/607#issuecomment-40143394.

teravest committed:
incorporated map background and consistent menus

teravest commented:
The map background is on webdev now.  I'm still working on correct verification for the passwords matching.

teravest committed:
fixed password validation

teravest committed:
add more validation

teravest commented:
Ok everything should be up on webdev now.

teravest committed:
fix issues with sess

teravest commented:
Alright I think this is ready for final review and lab wide testing.  @adamrp and @ElDeveloper can you check it over in the morning and let me know what you think?

ElDeveloper commented on database/procedures/ag_verify_password_change_code.sql:
```diff
@@ -0,0 +1,34 @@
+select
+case
+    when (current_time_ < k.PASS_RESET_TIME)
+   --when(current_time_ between (k.PASS_RESET_TIME - interval '2' hour) and k.PASS_RESET_TIME)
```
Should this line be removed here?

ElDeveloper commented on www/americangut/drawmap.psp:
```diff
@@ -0,0 +1,229 @@
+          featureType: "all",
+          elementType: "all",
+          stylers: [
+            { saturation: -100 } // <-- THIS
```
haha

ElDeveloper commented on www/americangut/drawmap.psp:
```diff
@@ -0,0 +1,229 @@
+%>
+
+
+<script type="text/javascript">
```
@teravest is there any of this JS code that can be centralized? i. e. is any of this used anywhere else?

teravest replied:
This was all copied out of index.psp,    I think in a future pr we can
update index.psp and register_user.psp to include drawmap.psp and then all
our mapping code will be in one place.

On Mon, Apr 14, 2014 at 4:09 PM, Yoshiki Vázquez Baeza <
notifications@github.com> wrote:

> In www/americangut/drawmap.psp:
> 
> > +**copyright** = "Copyright 2010, Qiime Web Analysis"
> > +**credits** = ["Emily TerAvest", "Adam Robbins-Pianka"]
> > +**license** = "GPL"
> > +**version** = "1.0.0.dev"
> > +**maintainer** = ["Emily TerAvest"]
> > +**email** = "emily.teraavest@colorado.edu"
> > +**status** = "Development"
> > +
> > +
> > +
> > +# Session variables
> > +sess = Session.Session(req)
> > +%>
> > +
> > +
> > +<script type="text/javascript">
> 
> @teravest https://github.com/teravest is there any of this JS code that
> can be centralized? i. e. is any of this used anywhere else?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/607/files#r11610944
> .

ElDeveloper replied:
Nix that, I just noticed that this is the centralized location.

ElDeveloper commented on www/americangut/drawmap.psp:
```diff
@@ -0,0 +1,229 @@
+    except:
+        req.write('<div class="header clearfix">\n')
+        req.write('<center><img id="logo" src="img/ag_logo.jpg">\n')
+        req.write('<br/>\n')
```
<br> does not need to be on it's own line, so I would recommend having this only in one or two lines (maybe).

adamrp replied:
Also, the tag should not have the slash -- should be <br>
On Apr 14, 2014 4:10 PM, "Yoshiki Vázquez Baeza" notifications@github.com
wrote:

> In www/americangut/drawmap.psp:
> 
> > +if 'map_command' not in sess:
> > -    # Write out the database of points
> >   +
> > -    result_list=[]
> >   +
> > -    # Get the map markers
> > -    ag_data_access = data_access_factory(ServerConfig.data_access_type, 'american_gut')
> >   +
> > -    # if database connection was unsuccessful, this is the first place it would
> > -    # manifest
> > -    try:
> > -        results = ag_data_access.getMapMarkers()
> > -    except:
> > -        req.write('<div class="header clearfix">\n')
> > -        req.write('<center><img id="logo" src="img/ag_logo.jpg">\n')
> > -        req.write('<br/>\n')
> 
> does not need to be on it's own line, so I would recommend having this
> only in one or two lines (maybe).
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/607/files#r11610972
> .

ElDeveloper commented on www/americangut/forgot_password.psp:
```diff
@@ -0,0 +1,229 @@
+        $('#resetpass_request').submit();
+}
+</script>
+<%@include file="loggedoutheader.psp"%>
```
niiiiiiiiiiiice

ElDeveloper commented on www/americangut/style/americangut.css:
```diff
@@ -144,6 +144,20 @@ div.registerwrapper {
 	text-align: left;
 }
 
+div.retrievekitwrapper {
```
These two classes, seem like exactly the same, can you rename these an change the elements that were referring to these classes so they refer to this new class?

teravest committed:
address comments

wasade commented on www/americangut/forgot_password.psp:
```diff
@@ -0,0 +1,229 @@
+    
+    var valid = true;
+    
+    if(!validateEmail(document.resetpass_request.email.value))
```
just curious, does `validateEmail` do any sql scrub?

teravest replied:
stored procedures are a defense against sql injection.
https://www.owasp.org/index.php/SQL_Injection_Prevention_Cheat_Sheet#Defense_Option_2:_Stored_Procedures
however thoughout the codebase anywhere we generate an sql statement we
should be using params like this
http://guyharrison.squarespace.com/blog/2010/1/17/best-practices-with-python-and-oracle.html
I tested this PR  the email validation wont let a space in to do something
like 'blah@blah.com; drop table test_table'
and in your comment below  because of the stored procedure 'kit_id; drop
table test_table; does not work either.

On Tue, Apr 15, 2014 at 11:16 AM, Daniel McDonald
notifications@github.comwrote:

> In www/americangut/forgot_password.psp:
> 
> > +from time import strftime
> > +from utils.mail import send_email, can_send_mail
> > +from string import letters, digits
> > +from urllib import quote, unquote
> > +
> > +%>
> > +<script type="text/javascript">
> > +function validatePassRequest() {
> > -    for(var i = 0; i < document.resetpass_request.length; i++)
> > -    {
> > -        document.resetpass_request[i].className = document.resetpass_request[i].className.replace(/(?:^|\s)highlight(?!\S)/ , '');
> > -    }
> >   +
> > -    var valid = true;
> >   +
> > -    if(!validateEmail(document.resetpass_request.email.value))
> 
> just curious, does validateEmail do any sql scrub?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/607/files#r11645696
> .

wasade commented on www/americangut/forgot_password.psp:
```diff
@@ -0,0 +1,229 @@
+    else:
+        email = unquote(form['email'])
+        supplied_kit_id = form['kit_id']
+        agDataAccess.ag_update_kit_password(supplied_kit_id, new_password)
```
same sql scrub question here, or are oracle stored procs protected from someone doing `"; drop table foo;"`?

wasade commented on www/americangut/retrieve_kitid.psp:
```diff
@@ -0,0 +1,102 @@
+                req.write("<br><br>")
+                req.write(MESSAGE)
+        else:
+            req.write('<p style="color:red;">This email address is not in our system</p>')
```
I've always been torn on this with systems when you click "forgot username" or the equiv. Basically, by providing this message, someone can test for whether a given person does or does not have an account. However, if we fail to indicate this, then it could easily lead to confusion as our participants sometimes have multiple email addresses and they do not always remember what they registered with

wasade commented:
This looks really awesome, sorry for the delay on review. Just a few comments

wasade commented:
Also, this is out of scope for this PR, but I realized looking through the AG data access code that every method creates a new connection instead of relying on an existing established connection in the object itself. That may be a piece of low hanging fruit for both a (very) minor performance gain and a reduction in boilerplate code across the object.

wasade commented:
When I click forgot kit ID, and enter an invalid email, it says the email is not found but does not give me the option to try entering another email.

teravest commented:
Do you want to create a new issue for that?

On Tue, Apr 15, 2014 at 11:27 AM, Daniel McDonald
notifications@github.comwrote:

> Also, this is out of scope for this PR, but I realized looking through the
> AG data access code that every method creates a new connection instead of
> relying on an existing established connection in the object itself. That
> may be a piece of low hanging fruit for both a (very) minor performance
> gain and a reduction in boilerplate code across the object.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/607#issuecomment-40509780
> .

wasade commented:
See #676

teravest committed:
adding form when invalid email

teravest commented:
fixed

On Tue, Apr 15, 2014 at 1:01 PM, Daniel McDonald
notifications@github.comwrote:

> When I click forgot kit ID, and enter an invalid email, it says the email
> is not found but does not give me the option to try entering another email.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/qiime/qiime_web_app/pull/607#issuecomment-40520598
> .

teravest committed:
make all emails in the database lowercase

adamrp commented:
So! Are we good to merge this now?

teravest commented:
I think this one is all set!

ElDeveloper commented:
:+1:

ElDeveloper commented:
This is awesome @teravest
